### DISCUSSION TITLE
Limiting max log record sizes produced by Zeek

### DISCUSSION THREAD

**Discussion by awelzel** [zeek/zeek#2797] [Projects]
## About

Provide mechanism to limit/cap the size of log records produced by Zeek. The main motivation is for protection from DoS either on the Zeek side due to producing overly large log record, or downstream systems due to overload or space exhaustion.

## Background

Log records produced by Zeek are currently not size limited in any way when rendered into JSON or TSV. For log records having a  "bounded representation" (int, addr, port, enum, bool, time, double, ...) this isn't an issue. A log record containing string, set or vector types, however is subject to grow unbounded (or minimally be of unreasonable size).

Scripting errors, analyzer errors, pathological traffic patterns or simply including non-sanitized input from analyzers into log fields can result in unintentionally producing enormous log record sizes.

## Downstream consumers

Downstream consumers or storage systems may impose limits on the size of records they allow. Either on the size of the whole log record, or on individual values/fields within a log record. If these limits are exceeded, data might be discarded or truncated and with luck reported through metrics.

Without limits, processing overhead increase on the Zeek side, but also on downstream consumers. There [have been issues around this](https://github.com/zeek/zeek/pull/2302). Once downstream consumers are unable to keep up and fall behind processing incoming logs, eventually something else will run out of buffer space cause data loss. There have also been issues by downstream systems rejecting too large messages, but producers not recognizing the error code and retrying endlessly.

### Example limits

Minimal survey of a few limits out there. Happy to hear of some more.

* Humio - 1mb event size: https://library.humio.com/falcon-logscale/limits.html
* MongoDB - 16 mb max document size: https://www.mongodb.com/docs/manual/reference/limits/#bson-documents
* filebeat log input - 10mb default max: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-log.html#_max_bytes_2
* Kafka message max bytes - default 1mb: https://kafka.apache.org/documentation/#brokerconfigs_message.max.bytes
* Loki max_label_value_length: <int> | default = 2048: https://grafana.com/docs/loki/latest/configuration/#limits_config

## Proposal

Introduce limits on the number of bytes allowed for individual string and container fields, as well as in total. Other field types are not relevant for capping the size with the assumption that they are small enough and may therefore not contribute significantly to the total size (with respect to generally large limits).

Different record types have different maximum sizes depending on their complexity, though are constrained by the "total" limits. A record type with four string log fields, for example, has twice the maximum allowed size than a record field with 4 string fields.

#### Summary
Introduce 4 new options, representing limits for individual fields and for the whole total of a record.
```
Log::limit_field_string_bytes         # Recommended default: 2KB
Log::limit_field_container_elements   # Recommended default: 50 (?)
Log::limit_total_string_bytes         # Recommended default: 250KB (? -250 KB of string data is large.
Log::limit_total_container_elements   # Recommended default 500 (?)
```

Extension/customization could be done on a per-stream/per-filter options overriding the global ones: For those use-cases where logging a 1MB string is on purpose.

For the actual processing, treat each record separately. When it's going through `Log::write()` and is converted to a `threading::Val` withing the logging Manager:
* Keep track of the total number of container elements and total number of string bytes (after capping)
* Cap string fields to `limit_field_string_bytes`
* Cap containers (vector, table, set) to `limit_field_container_elements`
* For other elements, process them, don't keep track.

* If limit_total_string_bytes is reached:
  * Render all strings as empty string "" unless within containers
  * Convert any containers holding strings to empty containers, if currently in such a container, stop processing it.
* If limit_total_container_elements is reached:
  * Convert any container of any type to empty containers
  * If currently in a container, return.

Essentially, there are two stages / configuration levels
  * Per field truncation
  * Emergency truncation once the "total" limits are hit.

I believe this could actually be implemented just with log policy hooks, too. It might not be the most efficient, however.

### Examples

The following record with the proposed default values would be bounded at 2*2KB + "few bytes for time" that aren't tracked.
```
type Info: record {
  t: time &log;
  s: string &log;
  t: string &log;
}
```

The following record may reach a total of 2KB + 50 * 2KB string bytes. The vector of enum values may reach a maximum size of 50 elements and then be capped, the enums will not be counted towards string bytes (even with large enum names, it's difficult to cause significant sizes).
```
type Info: record {
  s: string
  v: vector of string &log;
  e: vector of SomeEnum;
}
```

### Out-of-band: telemetry / metrics

We can expose per `record$field` occurrences of truncation as telemetry. Number of times a field was truncated and possibly the sum of bytes that were truncated. Further, "emergency" truncation can be tracked via per Record type metrics.

### reporter.log

For a combination of record type and field name, we could do a "warn once" or "warn once in a while" through reporter.log if truncation is happening so that it'll be visible on existing/established channels.

#### New Fields

* Adding new fields to log records to indicate truncation. Tricky, changes log schema.

#### Use N+1 limits
* From @bbannier
 
> Another option would be to let users configure some maximum length N, but then emit fields of length N+1 in case of overflow (we could terminate with a marker byte, but not strictly necessary). This would allow downstream consumers to detect the truncation without us having to change the schema.

I've been mulling over this and I'd personally pass on the N+1. The limits are intended to be set way larger than what's expected. If the output happens to hit the limit (or anywhere near 1% of the limit) it's likely already unexpectedly large. If it hits the limit, in very likely got truncated.

My take is that using strings of length N+1 will only bring up questions, require additional documentation, and I doubt downstream consumers would actually implement something clever with it (and then they could also implement the cleverness using the actual limit). On the other hand, a user testing something, and setting a limit/cap to 100 and then seeing strings of length 101 will think the implementation is buggy (I would).

**Comment by bbannier** [comment]
> > Another option would be to let users configure some maximum length N, but then emit fields of length N+1 in case of overflow (we could terminate with a marker byte, but not strictly necessary). This would allow downstream consumers to detect the truncation without us having to change the schema.
> 
> I've been mulling over this and I'd personally pass on the N+1. The limits are intended to be set way larger than what's expected. If the output happens to hit the limit (or anywhere near 1% of the limit) it's likely already unexpectedly large. If it hits the limit, in very likely got truncated.
> 
> My take is that using strings of length N+1 will only bring up questions, require additional documentation, and I doubt downstream consumers would actually implement something clever with it (and then they could also implement the cleverness using the actual limit). On the other hand, a user testing something, and setting a limit/cap to 100 and then seeing strings of length 101 will think the implementation is buggy (I would).

When talking about logging individual strings this approach aims to transport information on whether truncation occurred to downstream consumers without breaking the schema; this is only useful if consumers actually make use of that information (e.g., one cannot in general compare two truncated strings for equality). Since it makes things more complicate, I'd be curious to hear from users whether that is actually something they care about.

If we'd go that route we would need to come up with ways to also preserve such truncation information for containers, records, or whole log lines.

**Comment by timwoj** [comment]
I finally got a minute to read through this, and I like what you've laid out. I think the only thing I have a comment on is this part:

> Extension/customization could be done on a per-stream/per-filter options overriding the global ones: For those use-cases where logging a 1MB string is on purpose.

So you're thinking a set of options in the global Log namespace, but also separate options down at the stream/filter level that default to the global ones?

**Comment by awelzel** [comment]
> I finally got a minute to read through this, and I like what you've laid out. I think the only thing I have a comment on is this part:

It's been a bit. I read it again and I think I still like it, too.

> So you're thinking a set of options in the global Log namespace, but also separate options down at the stream/filter level that default to the global ones?

Yeah, but maybe just the global ones would be sufficient as a start? If more fine-grained is wanted, today I'd say the stream level is sufficient. 

PS:

> A record type with four string log fields, for example, has twice the maximum allowed size than a record field with 4 string fields.

Not sure what this was about :-) I think that "4" should've been "two".

**Comment by timwoj** [comment]
> ```
> Log::limit_field_string_bytes         # Recommended default: 2KB
> Log::limit_field_container_elements   # Recommended default: 50 (?)
> ```
> 

I ran into a few btests, specifically some SSL-related ones, that required increasing these limits. 4KB and 100 as defaults make the btests happy again without setting anything specific in the tests themselves.

> Reporting of truncation

My only concern here is that we'll silently truncate something in some new code/btest and that gets codified into the output from the test. Maybe if you're running in a testing context (i.e. `::running_under_test` is set to `T`), this should be noisier and possibly a fatal error.

**Comment by ckreibich** [comment]
I like this approach, too, I think it's actually very close to conversations we had in the past about this at Corelight. A few thoughts:

- Seems important that there's a per-filter / writer-specific way to customize this, since it can depend a lot on a given log writer how the log entry looks once written, and what a corresponding ingester expects.
- I wouldn't mark up the data itself as truncated. But I like the idea of supporting rendering such a "bit"  if somebody would like to, perhaps similar to how we can currently use extension fields to indicate other metadata. Feels a bit chicken-and-egg, but perhaps there's a clever way to provide that info sufficiently late but prior to writing the final output.
- Regarding reporting the occurrence, thumbs up from me for telemetry & reporter.log — they each have advantages: telemetry would be seamless and should work well with labels for log & field, and the latter can be specific about when truncation happened, which can help place it in an investigation. The biggie to keep in mind is that these things should be very rare.

We can nit-pick the thresholds — 2KB strikes me as a bit short for a string cap, but that's a detail.

**Comment by timwoj** [comment]
I created an issue from this discussion so that it can be linked to work done for it.
